src/mainwin.cc

```diff
@@ -2013,6 +2017,166 @@ void MainWindow::csgRender()
 	compileEnded();
 }
 
+
+void MainWindow::action3DPrint()
+{
+	//Keeps track of how many times we've exported and tries to create slightly unique filenames.
+	//Not mission critical, since non-unique file names are fine for the API, just harder to 
+	//differentiate between in customer support later.
+	static unsigned int printCounter=0;
+	
+	setCurrentOutput();
+	PRINT("3D Printing...");
+	
+	unsigned int dim = 3;
+	
+	QUrl partUrl;
+	
+	//Where we hold our temporary stl file name:
+	char tempStlFileName[L_tmpnam];
+	
+	setCurrentOutput();
+
+	//Make sure we can export:
+	if (! canExport(dim))
+	{
+		PRINT("Cannot 3D Print due to errors.");
+		return;
+    }
+	
+	//Create a temporary file name valid on all systems:
+	QTemporaryFile exportFile;
+	//Open the file so we can get the name:
+	if ( ! exportFile.open())
+	{
+		PRINT("Could not open temporary file.");
+		return;
+	}
+	QString exportFilename=exportFile.fileName();
+	
+	//Render the stl to a temporary file:
+	exportFileByName(this->root_geom, FileFormat::STL,
+		exportFilename.toLocal8Bit().constData(),
+		exportFilename.toUtf8());
+	
+	//Create a name that the order process will use to refer to the file. Base it off of the
+	//project name
+	QString userFacingName="unsaved.stl";
+	if (!this->fileName.isEmpty()) {
+		QString fileBasename = QFileInfo(this->fileName).baseName();
+		userFacingName = fileBasename.append("_").append(QString::number(printCounter++)).append(".stl");
+	}
+	
+	//Upload the file to the 3D Printing server and get the corresponding url to see it.
+	//The result is put in partUrl.
+	bool uploadWorked=uploadStlAndGetPartUrl(exportFilename, userFacingName, partUrl);
+	setCurrentOutput();
+
+	//Check the result:
+	if (! uploadWorked)
+	{
+		PRINT("An error occured while contacting the 3D Print API.");
+		return;
+	}
+	
+	//PRINT("partUrl:");
+	//PRINT(partUrl);
```

Comment:
Stray comments

Reply:
Hi Marius.  I cleaned up most of the print statements I had in the code before I submitted the pull request.  The ones remaining I thought would be useful if I'm ever in here debugging some issue we have.  They print key things like the json text we're sending, the json text we get back, and the url the function returns. All of that is important at a very top level for debugging.  Can we keep it in the codebase for future use so I don't end up having to write it again when I'm debugging something?

Reply:
For personal debug code, I usually just leave them in my private branch. Otherwise, the entire codebase would be littered with commented out debug code.
If the debug output is generally useful, you can use the PRINTDB macro, which will allow it to be compiled into the app and enabled at runtime.

Reply:
`#ifdef DEBUG` or `#if 0 // debug` could be options for such debug code.
Both can be found in our code base.

Reply:
Yes, for generally useful debug code, the best option is `PRINTD()` and `PRINTDB()`:

https://github.com/openscad/openscad/blob/44643571800e79e1c031a807e64052508e6e1b1f/src/printutils.h#L40-L52